src/main/java/org/openstreetmap/atlas/geography/atlas/raw/slicing/RawAtlasCountrySlicer.java

```diff
@@ -1,102 +1,311 @@
+ * @author samg
  */
 public class RawAtlasCountrySlicer
 {
+
+    // Logging constants
+    private static final String STARTED_TASK_MESSAGE = "Started {} for Shard {}";
+    private static final String COMPLETED_TASK_MESSAGE = "Finished {} for Shard {} in {}";
+    private static final String DYNAMIC_ATLAS_CREATION_TASK = "Dynamic Atlas Creation";
+
     private static final Logger logger = LoggerFactory.getLogger(RawAtlasCountrySlicer.class);
 
+    private final Sharding sharding;
+
+    private final Function<Shard, Optional<Atlas>> atlasFetcher;
+
     // The countries to be sliced with
     private final CountryBoundaryMap countryBoundaryMap;
 
     // The boundaries used for slicing
     private final Set<String> countries = new HashSet<>();
 
+    private final List<Shard> loadedShards = new ArrayList<>();
+
+    // Bring in all points -- this is important for proper sectioning later downstream, otherwise we
+    // will bring in geometry definitions with no underlying points
+    private static final Predicate<AtlasEntity> pointPredicate = entity -> entity instanceof Point;
+
+    // Bring in all relations that are tagged as Water or Coastline
+    private static final Predicate<AtlasEntity> relationPredicate = entity -> entity.getType()
+            .equals(ItemType.RELATION)
+            && Validators.isOfType(entity, NaturalTag.class, NaturalTag.WATER,
+                    NaturalTag.COASTLINE);
+
+    // Dynamic expansion filter will be a combination of the water relations and points
+    private static final Predicate<AtlasEntity> dynamicAtlasExpansionFilter = entity -> pointPredicate
+            .test(entity) || relationPredicate.test(entity);
+
     /**
-     * Slices against the given country and boundary map. Note: Will assume the entire world to be
-     * the target MultiPolygon to use.
+     * The default constructor for the old, pre-water relation pipeline-- this method will slice
+     * without dynamic expansion along water relations. Left in for legacy compatibility.
      *
      * @param countries
-     *            The Set of countries to be sliced against
+     *            the list of countries to slice
      * @param countryBoundaryMap
-     *            The {@link CountryBoundaryMap} to use when slicing
+     *            The country boundary map
      */
     public RawAtlasCountrySlicer(final Set<String> countries,
             final CountryBoundaryMap countryBoundaryMap)
     {
         this.countries.addAll(countries);
         this.countryBoundaryMap = countryBoundaryMap;
+        this.sharding = null;
+        this.atlasFetcher = null;
+    }
+
+    /**
+     * Updated constructor for the water relation pipeline-- this method will use the Atlas fetcher
+     * function and sharding tree to dynamically expand on all water relations so they can be sliced
+     * appropriately.
+     *
+     * @param country
+     *            The list of countries to slice. Should really just be one country?
```

Comment:
Should really just be one country should be removed?

Reply:
hmm, not sure, this wasn't a very clear way of expressing this. with the decoupling of point-and-line slicing and relation slicing, it seems like the most well-supported code paths are to country slice each shard one country at a time. that's not to say multiple country slicing doesn't work, but rather to say that we should probably think about deprecating it?